lib/temporal/workflow/state_manager.rb

```diff
@@ -83,43 +89,98 @@ def apply(history_window)
         @replay = history_window.replay?
         @local_time = history_window.local_time
         @last_event_id = history_window.last_event_id
+        history_window.sdk_flags.each { |flag| sdk_flags.add(flag) }
 
-        # handle markers first since their data is needed for processing events
-        history_window.markers.each do |event|
+        order_events(history_window.events).each do |event|
           apply_event(event)
         end
+      end
 
-        history_window.events.each do |event|
-          apply_event(event)
+      def self.event_order(event, signals_first)
+        if event.type == 'MARKER_RECORDED'
+          # markers always come first
+          0
+        elsif event.type == 'WORKFLOW_EXECUTION_STARTED'
+          # This always comes next if present
+          1
+        elsif signals_first && signal_event?(event)
+          # signals come next if we are in signals first mode
+          2
+        else
```

Comment:
You also need to sort _queries_ in a distinct category, after everything else.

Reply:
I don't believe there is any code that handles query history events specifically. Actually, I don't see any history events that have to do with queries that would be handled here. All query handling code lives at the task processor level, and doesn't involve anything in the state manager as far as I know.

@dwillett What do you think?